server/container/util/src/main/java/org/apache/james/util/ReactorUtils.java

```diff
@@ -100,4 +109,61 @@ public int read() {
     private static int byteToInt(ByteBuffer buffer) {
         return buffer.get() & 0xff;
     }
+
+    public static Consumer<Signal<?>> logOnError(Consumer<Throwable> errorLogStatement) {
+        return signal -> {
+            if (!signal.isOnError()) {
+                return;
+            }
+            Optional<MDCBuilder> maybeMDC = retrieveMDCBuilder(signal);
+
+            if (maybeMDC.isPresent()) {
+                try {
+                    try (Closeable mdc = maybeMDC.get().build()) {
+                        errorLogStatement.accept(signal.getThrowable());
+                    }
+                } catch (IOException e) {
+                    throw new RuntimeException(e);
+                }
+            } else {
+                errorLogStatement.accept(signal.getThrowable());
+            }
+        };
+    }
+
+    public static Consumer<Signal<?>> log(Runnable logStatement) {
+        return signal -> {
+            Optional<MDCBuilder> maybeMDC = retrieveMDCBuilder(signal);
+
+            if (maybeMDC.isPresent()) {
+                try {
+                    try (Closeable mdc = maybeMDC.get().build()) {
+                        logStatement.run();
+                    }
+                } catch (IOException e) {
+                    throw new RuntimeException(e);
+                }
+            } else {
+                logStatement.run();
+            }
+        };
+    }
+
+    public static Context context(String keySuffix, MDCBuilder mdcBuilder) {
+        return Context.of(mdcKey(keySuffix), mdcBuilder);
+    }
+
+    public static String mdcKey(String value) {
+        return MDC_KEY_PREFIX + value;
+    }
+
+    private static Optional<MDCBuilder> retrieveMDCBuilder(Signal<?> signal) {
```

Comment:
we should return an empty MDCBuilder when there's not entry to avoid handling emptiness